### PR TITLE
Ray task proposal

### PR DESCRIPTION
**Pull request by pingsutw**
# TL;DR
https://docs.google.com/document/d/1-695lxz8a-GFz4cFamGkF1NhspKMkDGIDdkY9EDLMz8/edit#

## Type
 - [ ] Bug Fix
 - [x] Feature
 - [ ] Plugin

## Are all requirements met?

 - [x] Code completed
 - [x] Smoke tested
 - [ ] Unit tests added
 - [x] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/2641

## Follow-up issue
_NA_

### CODE REVIEW

File: protos/flyteidl/core/resource.proto [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+// Cluster resource that task and workflow will run on
+message Resource {
+  oneof value {
+    RayCluster ray = 1;
+  }
+}
+
+
+// Define Ray cluster spec
+message RayCluster {
```
Comment (wild-endeavor) [inline protos/flyteidl/core/resource.proto:18]: should we move the Ray specific messages into a separate file?  If not, should we add "Ray" to the names?  I know renaming is annoying cuz of all the generated/model code you have to

File: protos/flyteidl/plugins/ray.proto [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+  // The worker group configurations
+  repeated WorkerGroupSpec worker_group_spec = 2;
+}
+
+message HeadGroupSpec {
+  // Optional. The ray start params of head node group
+  map<string, string> ray_start_params = 1;
+}
+
+message WorkerGroupSpec {
```
Comment (daikeshi) [inline protos/flyteidl/plugins/ray.proto:42]: Could we also add `service_account_name` to `HeadGroupSpec` and `WorkerGroupSpec`? This is useful for folks when they use a specific service account with necessary permissions configured to access cloud resources.

Comment (pingsutw) [reply protos/flyteidl/plugins/ray.proto:42]: We can configure the service account or IAM role on flyteconsole or during registration.
`pyflyte register --service-account default ...`

Screen Shot 2022-07-29 at 3 10 19 AM

Comment (pingsutw) [reply protos/flyteidl/plugins/ray.proto:42]: Actually, some task level configs (like resource, secret, service account, IAM) we already have in Flyte. Therefore, we don't need to define it here. https://github.com/flyteorg/flyteidl/blob/e749b82a70027d3364f4e5b98ce2cfbffb4eef54/protos/flyteidl/admin/launch_plan.proto#L96-L107

File: protos/flyteidl/plugins/ray.proto [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+// RayJobSpec defines the desired state of RayJob
+message RayJob {
+  // RayClusterSpec is the cluster template to run the job
+  RayCluster ray_cluster = 1;
+  // runtime_env is base64 encoded.
+  // Ray runtime environments: https://docs.ray.io/en/latest/ray-core/handling-dependencies.html#runtime-environments
+  string runtime_env = 2;
+  // shutdown_after_job_finishes will determine whether to delete the ray cluster once rayJob succeed or failed.
+  bool shutdown_after_job_finishes = 3;
```
Comment (kumare3) [inline protos/flyteidl/plugins/ray.proto:17]: doesnt this seem odd to be encoded into the request? Should this be implicit?

File: protos/flyteidl/plugins/ray.proto (hunk 2 above)
Comment (wild-endeavor) [reply protos/flyteidl/plugins/ray.proto:42]: @daikeshi does this sound reasonable to you?

**Review by wild-endeavor — DISMISSED** [review]

File: protos/flyteidl/plugins/ray.proto (hunk 3 above)
Comment (pingsutw) [reply protos/flyteidl/plugins/ray.proto]: removed it, I think we can move it to propeller's config

**Review by wild-endeavor — DISMISSED** [review]

**Review by wild-endeavor — APPROVED** [review]

**Review by kumare3 — APPROVED** [review]

**Pull request merged**